runtime/dancebox/src/lib.rs

```diff
@@ -811,54 +816,127 @@ impl RemoveInvulnerables<CollatorId> for RemoveInvulnerablesImpl {
+        } else {
+            blocks_per_session
+        };
+        // Check if the container chain has enough credits for producing blocks
+        let free_block_credits =
+            pallet_services_payment::BlockProductionCredits::<Runtime>::get(para_id)
+                .unwrap_or_default();
+        let remaining_block_credits = block_credits_needed.saturating_sub(free_block_credits);
+        let (block_production_costs, _) =
+            <Runtime as pallet_services_payment::Config>::ProvideBlockProductionCost::block_cost(
+                &para_id,
+            );
+        // Check if we can withdraw
+        let remaining_block_credits_to_pay =
+            u128::from(remaining_block_credits).saturating_mul(block_production_costs);
+        let remaining_to_pay = remaining_block_credits_to_pay;
+        // This should take into account whether we tank goes below ED
+        // The true refers to keepAlive
+        Balances::can_withdraw(
+            &pallet_services_payment::Pallet::<Runtime>::parachain_tank(para_id),
+            remaining_to_pay,
+        )
+        .into_result(true)?;
+        // TODO: Have proper weight
+        Ok(Weight::zero())
+    }
+}
 
-            let remaining_to_pay = remaining_block_credits_to_pay.saturating_add(remaining_session_credits_to_pay).saturating_add(max_tip);
+impl<AC> RemoveParaIdsWithNoCredits<BalanceOf<Runtime>, AC> for RemoveParaIdsWithNoCreditsImpl {
+    fn pre_assignment_remove_para_ids_with_no_credits(
+        para_ids: &mut Vec<ParaId>,
+        currently_assigned: &BTreeSet<ParaId>,
+    ) {
+        let blocks_per_session = Period::get();
+        para_ids.retain(|para_id| {
+            with_transaction(|| {
+                let max_tip =
+                    pallet_services_payment::MaxTip::<Runtime>::get(para_id).unwrap_or_default();
+                TransactionOutcome::Rollback(Self::charge_para_ids_internal(
+                    blocks_per_session,
+                    *para_id,
+                    currently_assigned,
+                    &Some(max_tip),
+                ))
+            })
+            .is_ok()
+        });
+    }
 
-            // This should take into account whether we tank goes below ED
-            // The true refers to keepAlive
-            Balances::can_withdraw(&pallet_services_payment::Pallet::<Runtime>::parachain_tank(*para_id), remaining_to_pay).into_result(true).is_ok()
+    fn post_assignment_remove_para_ids_with_no_credits(
+        current_assigned: &BTreeSet<ParaId>,
+        new_assigned: &mut BTreeMap<ParaId, Vec<AC>>,
+        maybe_tip: &Option<BalanceOf<Runtime>>,
+    ) -> Weight {
+        let blocks_per_session = Period::get();
+        let mut total_weight = Weight::zero();
+        new_assigned.retain(|&para_id, collators| {
+            // Short-circuit in case collators are empty
+            if collators.is_empty() {
+                return true;
+            }
+            with_storage_layer(|| {
```

Comment:
what is `with_storage_layer`¿?

Reply:
So, `with_storage_layer` is similar to `with_transaction` but it automatically commits the storage changes into parent overlay if the closure returns `Ok` and do the rollback otherwise. 

Parity added this function since it is more convenient to use. (i.e you don't have to manually return `Ok` or `Err` for commit and rollback respectively.

Reply:
Are there any performance concerns when running that function inside a loop? I guess not?

Reply:
Fairly low cost. See here: https://github.com/paritytech/substrate/pull/10809#issuecomment-1078343281

Reply:
*disregard*